x-pack/legacy/plugins/actions/server/builtin_action_types/webhook.ts

```diff
@@ -4,65 +4,94 @@
-
 const HeadersSchema = schema.recordOf(schema.string(), schema.string());
-
-const ConfigSchema = schema.object({
+const configSchemaProps = {
   url: schema.string(),
   method: schema.oneOf([schema.literal(WebhookMethods.POST), schema.literal(WebhookMethods.PUT)], {
     defaultValue: WebhookMethods.POST,
   }),
   headers: nullableType(HeadersSchema),
-});
+};
+const ConfigSchema = schema.object(configSchemaProps);
+type ActionTypeConfigType = TypeOf<typeof ConfigSchema>;
 
 // secrets definition
-export type ActionTypeSecretsType = TypeOf<typeof SecretsSchema>;
+type ActionTypeSecretsType = TypeOf<typeof SecretsSchema>;
 const SecretsSchema = schema.object({
   user: schema.string(),
   password: schema.string(),
 });
 
 // params definition
-export type ActionParamsType = TypeOf<typeof ParamsSchema>;
+type ActionParamsType = TypeOf<typeof ParamsSchema>;
 const ParamsSchema = schema.object({
   body: schema.maybe(schema.string()),
 });
 
 // action type definition
-export const actionType: ActionType = {
-  id: '.webhook',
-  name: 'webhook',
-  validate: {
-    config: ConfigSchema,
-    secrets: SecretsSchema,
-    params: ParamsSchema,
-  },
-  executor,
-};
+export function getActionType(configurationUtilities: ActionsConfigurationUtilities): ActionType {
+  return {
+    id: '.webhook',
+    name: 'webhook',
+    validate: {
+      config: schema.object(configSchemaProps, {
+        validate: curry(valdiateActionTypeConfig)(configurationUtilities),
+      }),
+      secrets: SecretsSchema,
+      params: ParamsSchema,
+    },
+    executor: curry(executor)(configurationUtilities),
+  };
+}
+
+function valdiateActionTypeConfig(
+  configurationUtilities: ActionsConfigurationUtilities,
+  configObject: ActionTypeConfigType
+) {
+  const { url } = configObject;
+  const whitelistValidation = configurationUtilities.isWhitelistedUri(url);
+  if (whitelistValidation instanceof NotWhitelistedError) {
+    return i18n.translate('xpack.actions.builtin.webhook.webhookConfigurationError', {
+      defaultMessage: 'error configuring webhook action: {message}',
+      values: {
+        message: whitelistValidation.message,
+      },
+    });
+  }
+}
 
 // action executor
-async function executor(execOptions: ActionTypeExecutorOptions): Promise<ActionTypeExecutorResult> {
+export async function executor(
+  configurationUtilities: ActionsConfigurationUtilities,
+  execOptions: ActionTypeExecutorOptions
+): Promise<ActionTypeExecutorResult> {
   const log = (level: string, msg: string) =>
     execOptions.services.log([level, 'actions', 'webhook'], msg);
 
   const id = execOptions.id;
   const { method, url, headers = {} } = execOptions.config as ActionTypeConfigType;
   const { user: username, password } = execOptions.secrets as ActionTypeSecretsType;
   const { body: data } = execOptions.params as ActionParamsType;
+  const whitelistValidation = configurationUtilities.isWhitelistedUri(url);
```

Comment:
I thought we re-ran the validators (all of them?) before invoking the executor?  Yeah, I think so, so this explicit check shouldn't be needed, and I think you'd never get into this method, and have the subsequent code run. https://github.com/elastic/kibana/blob/49fdd24f80fb46af2967fd5c705e4387b63e6526/x-pack/legacy/plugins/actions/server/lib/execute.ts#L50-L56

Reply:
Yup, I've double checked and this is true, but somewhat implicit as it requires prior knowledge of the wiring.
I've removed the redundant validation, but feel this is something worth revisiting in the future... perhaps there's a better way to model this so it's clearer.